### DISCUSSION TITLE
Memory and storage consumption

### DISCUSSION THREAD

**Discussion by vacavaca** [pocketbase/pocketbase#1224] [Q&A] [answered]
In a standard docker setup under a small load (0.1 - 1 rps I think, less than 100 records in db) pocketbase leaks memory (see graph) and produces a lot of logs (around 1GB / daily).

My questions are:

1) What can cause memory to leak in such a way? my server (2CPU/2GB droplet) goes out of memory twice a day. I've checked my other containers, all of them function properly.
2) Is there a way to write logs to stdout or to disable them completely? Under this load the logs page in the UI just doesn't work, and in my opinion in a docker setup it would be much more convinient to have logs in stdout and then route them to loki / journald / etc. similarly to other apps.
3) How to safely cleanup the logs db without restarting pocketbase?

<img width="1968" alt="Screenshot 2022-12-09 at 20 02 29" src="https://user-images.githubusercontent.com/19508396/206756697-0cc09bc1-7965-47d1-bc0d-5e82ce6928fd.png">

**Comment by ganigeorgiev** [comment, marked as answer]
What is the left side in the above graph - bytes?

Also, what version of PocketBase are you using?

You can disable the logs from the "Admin UI > Settings > Application > Logs max days retention" and set the field to `0` (I'll consider in the future to make them more granular).

For the memory consumption - there is a lot of room for improvements and allocation optimizations will be done in the future, but the last time I checked I couldn't find any significant memory growth that would indicate memory leak (eg. forgotten descriptors, uncapped cache, etc.). You can check https://github.com/pocketbase/pocketbase/issues/969#issuecomment-1308520247.

**Reply by vacavaca** [reply]
Thanks for the quick reply!

Yea, the left side of the graph is bytes. Can't check the exact version right now will check it later when I can get my hands to the server.

I have 7 days retention now. Maybe it's worth trying 1 day or so, thanks for pointing that out. I've missed that field.

**Reply by ganigeorgiev** [reply]
I suspect that the high memory consumption is due to the logs write operation. I'll try to stress test it again and will see if I can optimize it a little (or at least reduce the max allowed open connections to the logs db).

You can try temporarily disabling the logs by setting the field to `0` and leave it for a while to see if there will be any difference.

**Reply by ganigeorgiev** [reply]
@vacavaca Another workaround that you can try is to set the [`GOMEMLIMIT` env variable](https://pkg.go.dev/runtime@master#hdr-Environment_Variables) (eg. `GOMEMLIMIT=500MiB`).

It is a soft limit and will not stop the memory grow on high load but it will try to keep the idle app used memory within the set value.

_Note that this variable is available only with Go 1.19. The prebuilt executables already use 1.19, but you'll have to update your go version if you are embedding PocketBase in your Go app._

**Reply by vacavaca** [reply]
Thanks for the suggestion will try it a bit later.
I upgraded Pocketbase from 0.8.0 to 0.9.2 last night, but still experiencing problems. The memory graph is on the screenshot.

A bit more details about my setup:
- 50 - 70 records in the DB overall, light read load, basically zero write load. I'll try to get exact metrics tonight.
- Debian 11, linux 5.10.0-19-amd64
- Pocketbase is downloaded from GitHub distribution, like so:

```dockerfile
ENV BUILDX_ARCH="${TARGETOS:-linux}_${TARGETARCH:-amd64}"
RUN wget https://github.com/pocketbase/pocketbase/releases/download/v${VERSION}/pocketbase_${VERSION}_${BUILDX_ARCH}.zip \
    && unzip pocketbase_${VERSION}_${BUILDX_ARCH}.zip \
    && chmod +x /pocketbase
```
<img width="1990" alt="Screenshot 2022-12-10 at 22 54 50" src="https://user-images.githubusercontent.com/19508396/206873026-d51912bb-1b65-447d-b840-69e802b5d3b9.png">

**UPD**: Disabled logs as you suggested, monitoring now

**Reply by ganigeorgiev** [reply]
Enabling the `GOMEMLIMIT=300MiB` env variable should also help with the issue because it will force the GC to run more often and releasing any unneeded preserved memory.

There are several issues in the go repo reporting that in bursty/spiky scenarios the GC is not always scheduled in a timely manner that could cause the OOM errors (eg. https://github.com/golang/go/issues/42805, https://github.com/golang/go/issues/42430 and probably others). This is one of the reason why the soft limit was introduced (it is very similar concept to what you have in jvm for example).

As mentioned previously, memory allocation will be improved in the near future (_I've already pushed in master a [small fix related to the datetime parsing](https://github.com/pocketbase/pocketbase/commit/68a9782c03b523a19287df884ffadc8caf25e708) that was causing 20% of the overall heap; other minor improvements will also be applied soon_), but as mentioned in the findings in https://github.com/pocketbase/pocketbase/issues/969#issuecomment-1308520247 this is not a memory leak, just not very efficient allocations. Memory leaks in GC languages are rare and usually happens from forgotten/unreleased resources (eg. open file descriptors without close), unconstrained cache (eg. evergrowing map), etc. and I wasn't able to find such issues in the PocketBase source.

**Reply by vacavaca** [reply]
I disabled logs entirely, memory usage got back to normal, low, and flat graph. I'm not perfectly sure about the cause of it. I did some other unrelated maintenance, so there might be several effects at play here, nonetheless disabling logs works, so maybe it's worth investigating.

**Reply by ganigeorgiev** [reply]
@vacavaca It is probably because of the max open connections. It is already restricted in `master` alongside other minor improvements. The json serialization currently takes most of the time and it is responsible for most of the allocations due to the use of reflection. Manual/custom json serializer for the records will be available in v0.10 or v0.11.

**Reply by ganigeorgiev** [reply]
@vacavaca _Just to let you know that I've just release [PocketBase v0.10.0](https://github.com/pocketbase/pocketbase/releases/tag/v0.10.0) and you should expect lower memory usage even with enabled logs._
